src/undo_log.rs

```diff
@@ -0,0 +1,225 @@
+//! Module which contains the snapshot/rollback functionality of the `ena` data structures.
+//!
+//! For most usecases this is just an internal implementation detail. However if many `ena`
+//! data structures are used snapshotted simultaneously it is possible to use
+//! `UnificationTableStorage`/`SnapshotVecStorage` instead and use a custom `UndoLogs<T>`
+//! type capable of recording the actions of all used data structures.
+//!
+//! Since the `*Storage` variants do not have an undo log `with_log` must be called with the
+//! unified log before any mutating actions.
+
+/// A trait which allows actions (`T`) to be pushed which which allows the action to be undone at a
+/// later time if needed
+pub trait UndoLogs<T> {
```

Comment:
Can we add some bound to `T`?

Reply:
No, from the perspective of `UndoLogs` and `Rollback`, `T` is opaque. Only specific implementations of `Rollback` cares about the concrete type (which is just a set of instructions on how to rollback that specific type)

Reply:
OK, I thought that might be the case, let's document that.